### PR TITLE
Fix l7 redirects for host / world flows with AllowsLocalhost=always, HostAllowsWorld options

### PR DESCRIPTION
**Pull request by joestringer**
If the daemon options `AllowsLocalhost=always` (default with k8s) and/or
`LegacyHostAllowsWorld` are set, and any L7 rules are created that might
match traffic from host / world, then these L3 sources should be wildcarded
at L7 to ensure that proxies will not restrict the traffic from these sources.

This was previously not done, but instead the l4 redirects were removed
later during policymap entry generation. This prevented such traffic from
even reaching the proxy, which is counter to how the API for such a rule
combination is described.

This PR teaches the policy logic about these options so that L7 can be
appropriately wildcarded depending on the mode, and reinstates the
redirects when L7 policy is applied.

Fixes: #4401

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/4416)

### CODE REVIEW

**Comment by joestringer** [comment]
test-me-please

File: pkg/policy/l4.go
Comment (ianvernon) [inline pkg/policy/l4.go]: Update documentation to include information about the L3-overrides.

File: pkg/policy/l4.go
Comment (ianvernon) [inline pkg/policy/l4.go]: Is `labels[0]` guaranteed to exist? Wondering in the case of a potential deference of something that does not exist.

File: pkg/policy/l4Filter_test.go
Comment (ianvernon) [inline pkg/policy/l4Filter_test.go]: XXX?

File: pkg/policy/api/selector.go
Comment (ianvernon) [inline pkg/policy/api/selector.go]: Nit: can you change it to end with "that only map reserved labels (e.g. host, or world) to their corresponding endpoint selectors"  ? This more clearly specifies the mapping in the documentation.

File: pkg/policy/rule.go
Comment (ianvernon) [inline pkg/policy/rule.go]: Is this only necessary for ingress ? I'm guessing that it is only necessary for ingress based on the documentation for the CLI option for "always allow localhost":

https://github.com/cilium/cilium/blob/ad437706dd4ee26d704e05e3560d74ae7b2c113c/daemon/main.go#L310-L311

**Review by ianvernon — DISMISSED** [review]
Overall, change looks great. Glad it wasn't too invasive :)

File: pkg/policy/l4.go
Comment (joestringer) [reply pkg/policy/l4.go]: Its usage is described five lines lower, where it's used. I can move that comment up if you think that's more readable.

File: pkg/policy/l4.go
Comment (joestringer) [reply pkg/policy/l4.go]: Yes, it's created two functions up. I played around with a few options, maybe passing a `map[string]EndpointSelector` down would be more readable.

File: pkg/policy/l4Filter_test.go
Comment (joestringer) [reply pkg/policy/l4Filter_test.go]: Ah, I copy-pasted most of this test so didn't yet double-check everything. Let me take a look.

File: pkg/policy/rule.go
Comment (joestringer) [reply pkg/policy/rule.go]: Yes, it's only for ingress.

**Comment by joestringer** [comment]
Hit this timeout while contacting a service post-upgrade:

```
Stacktrace

/home/jenkins/workspace/Cilium-PR-Ginkgo-Tests-Validated/src/github.com/cilium/cilium/test/ginkgo-ext/scopes.go:373
Cannot curl app1-service
Expected command: kubectl exec -n default app2-3691078314-77l9v -- curl -s -D /dev/stderr --fail --connect-timeout 5 --max-time 5 http://app1-service.default.svc.cluster.local/public 
To succeed, but it failed:
Exitcode: 28 
Stdout:
 	 
Stderr:
 	 

/home/jenkins/workspace/Cilium-PR-Ginkgo-Tests-Validated/src/github.com/cilium/cilium/test/k8sT/Updates.go:112
```

Seems very similar to https://github.com/cilium/cilium/issues/4384, will retry the build.

[f0c5341e_K8sValidatedUpdates_Updating_Cilium_stable_to_master.zip](https://github.com/cilium/cilium/files/2083113/f0c5341e_K8sValidatedUpdates_Updating_Cilium_stable_to_master.zip)

**Comment by joestringer** [comment]
test-me-please

File: pkg/policy/l4.go
Comment (joestringer) [reply pkg/policy/l4.go]: Turns out that we don't need this hack any more, we can just pass a `[]api.EndpointSelector` around. I missed this when refactoring to simplify. I'll fix this up.

File: pkg/policy/l4.go
Comment (ianvernon) [reply pkg/policy/l4.go]: Please do, because otherwise the `godoc` will not specify what they are used for.

File: pkg/policy/l4.go
Comment (ianvernon) [inline pkg/policy/l4.go:194]: Awesome :D

File: pkg/policy/l4Filter_test.go
Comment (ianvernon) [inline pkg/policy/l4Filter_test.go:1205]: How is 'allow all' treated with precedence if there are more "specific" endpoint selectors? In this case, we are restricting for all endpoints, but then allowing all for the host selector. How will the datapath handle this?

File: pkg/policy/rule.go
Comment (ianvernon) [inline pkg/policy/rule.go:143]: Awesome comments throughout this PR!

File: pkg/policy/l4Filter_test.go
Comment (joestringer) [reply pkg/policy/l4Filter_test.go:1205]: I'm not exactly sure how the precedence is determined here, but I tested this PR manually and it had the behaviour I was looking for, ie the proxy now starts seeing all traffic from the host, though the host allows all at L3 (because there's an L7 policy wildcarding L3).

/cc @rlenglet @jrajahalme Is there a precedence defined between these? Or is the wildcardendpointselector handled in a special way?

Comment (joestringer) [reply pkg/policy/l4Filter_test.go:1205]: Walked through the `getDirectionNetworkPolicy()` logic and it seems like when this L4Filter logic is applied to a set of identities there, it properly resolves the set of rules for each identity based on the list of selectors here, ie the selectors here may overlap with one another, and if multilple select a pod, then the rules are added together to determine the allowed set. If one of the selectors allows all, then it will add "allow all" to "allow something more specific", and resolve to "allow all". That's why this works.

Comment (ianvernon) [reply pkg/policy/l4Filter_test.go:1205]: Awesome! :D

**Review by ianvernon — APPROVED** [review]

**Comment by ianvernon** [comment]
LGTM! Great work. Let's wait until the CI is stabilized before we get this merged.

**Comment by tgraf** [comment]
Merged via #4435

**Pull request closed without merging**
